geomstats/stiefel.py

```diff
@@ -0,0 +1,279 @@
+from geomstats.riemannian_metric import RiemannianMetric
+
+TOLERANCE = 1e-6
+EPSILON = 1e-6
+
+def matrix_f(sq_mat, f):
+    
+    sq_mat = gs.to_ndarray(sq_mat, to_ndim=3)
+    
+    [eigenvalues, vectors] = gs.linalg.eig(sq_mat)
+    
+    exp_eigenvalues = f(eigenvalues)
+    
+    aux = gs.einsum('ijk,ik->ijk', vectors, exp_eigenvalues)
+    exp_mat = gs.einsum('ijk,ikl->ijl', aux, gs.linalg.inv(vectors))
+    
+    return exp_mat.real
+
+class Stiefel(EmbeddedManifold):
+
+    def __init__(self, p, n):
+        self.p = p
+        self.n = n
+
+        self.dimension = int(p * n - (p * (p - 1) / 2))
+
+    def belongs(self, point, tolerance=TOLERANCE):
+        """
+        Evaluate if a point belongs to St(p,n),
+        i.e. if it is a p-frame in n-dimensional space,
+        and it is orthonormal.
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        (_, point_dim) = point.shape
+
+        if point_dim != (self.n, self.p):
+            return False
+        
+        point_norm = gs.norm(gs.dot(gs.transpose(point), point) - gs.eye(point_dim[1]))
+
+        return gs.less_equal(point_norm, tolerance)
+
+    def project(self):
+        pass
+
+    def random_uniform(self, n_samples=1):
+        """
+        Sample on St(p,n) with the uniform distribution.
+
+        If Z(p,n) ~ N(0,1), then St(p,n) ~ U, according to Haar measure:
+        St(p,n) := Z(Z^TZ)^{-1/2}
+        """
+        Z = gs.normal(size=(n_samples, self.n, self.p))
+        return gs.matmul(Z, gs.linalg.inv(matrix_f(gs.matmul(gs.transpose(Z, axes=(0, 2, 1)), Z), gs.sqrt)))
+
+class StiefelMetric(RiemannianMetric):
+
+    def __init__(self, dimension):
+        self.dimension = dimension
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point=None):
+        """
+        Compute the Frobenius inner product of tangent_vec_a and tangent_vec_b
+        at base_point using the canonical Riemannian metric on St(p,n).
+        """
+        tangent_vec_a = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        n_tangent_vecs_a, _, _ = tangent_vec_a.shape
+
+        tangent_vec_b = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        n_tangent_vecs_b, _, _ = tangent_vec_b.shape
+
+        assert(n_tangent_vecs_a == n_tangent_vecs_b)
+        
+        return gs.einsum("zij,zij->z", tangent_vec_a, tangent_vec_b)
+
+    def norm(self, tangent_vec_a, tangent_vec_b):
+        pass
+
+    def exp(self, tangent_vec, base_point):
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        n_tangent_vecs, _, _ = tangent_vec.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert(n_tangent_vecs == n_base_points
+            or n_tangent_vecs == 1
+            or n_base_points == 1)
+        
+        A = gs.matmul(gs.transpose(base_point, axes=(0, 2, 1)), tangent_vec)
```

Comment:
No uppercase in variable names + avoid single letter variable names (it's harder to understand what the variable represents and impossible to use the editor's "replace" feature).

https://www.python.org/dev/peps/pep-0008/#function-and-variable-names